crates/client-api/src/routes/database.rs

```diff
@@ -178,13 +151,146 @@ fn reducer_outcome_response(identity: &Identity, reducer: &str, outcome: Reducer
+        // then, well, the database is out of energy.
+        // Return 503 service unavailable.
+        ClientConnectedError::OutOfEnergy => (StatusCode::SERVICE_UNAVAILABLE, err.to_string()).into(),
+        // If `call_identity_connected` returns `Err(ReducerCall)`,
+        // something went wrong while invoking the `client_connected` reducer.
+        // I (pgoldman 2025-03-27) am not really sure how this would happen,
+        // but we returned 404 not found in this case prior to my editing this code,
+        // so I guess let's keep doing that.
+        ClientConnectedError::ReducerCall(e) => (StatusCode::NOT_FOUND, format!("{:#}", anyhow::anyhow!(e))).into(),
+        // If `call_identity_connected` returns `Err(DBError)`,
+        // then the module didn't define `client_connected`,
+        // but something went wrong when we tried to insert into `st_client`.
+        // That's weird and scary, so return 500 internal error.
+        ClientConnectedError::DBError(_) => (StatusCode::INTERNAL_SERVER_ERROR, err.to_string()).into(),
+    }
+}
+
+/// If `call_identity_disconnected` errors, something is very wrong:
+/// it means we tried to delete the `st_client` row but failed.
+///
+/// Note that `call_identity_disconnected` swallows errors from the `client_disconnected` reducer.
+/// Slap a 500 on it and pray.
+fn client_disconnected_error_to_response(err: ReducerCallError) -> ErrorResponse {
+    (StatusCode::INTERNAL_SERVER_ERROR, format!("{:#}", anyhow::anyhow!(err))).into()
+}
+
 #[derive(Debug, derive_more::From)]
 pub enum DBCallErr {
     HandlerError(ErrorResponse),
     NoSuchDatabase,
     InstanceNotScheduled,
 }
 
+#[derive(Deserialize)]
+pub struct ProcedureParams {
+    name_or_identity: NameOrIdentity,
+    procedure: String,
+}
+
+async fn procedure<S: ControlStateDelegate + NodeDelegate>(
+    State(worker_ctx): State<S>,
+    Extension(auth): Extension<SpacetimeAuth>,
+    Path(ProcedureParams {
+        name_or_identity,
+        procedure,
+    }): Path<ProcedureParams>,
+    TypedHeader(content_type): TypedHeader<headers::ContentType>,
+    ByteStringBody(body): ByteStringBody,
+) -> axum::response::Result<impl IntoResponse> {
+    if content_type != headers::ContentType::json() {
+        return Err(axum::extract::rejection::MissingJsonContentType::default().into());
+    }
+    let caller_identity = auth.claims.identity;
+
+    let args = FunctionArgs::Json(body);
+
+    let db_identity = name_or_identity.resolve(&worker_ctx).await?;
+    let database = worker_ctx_find_database(&worker_ctx, &db_identity)
+        .await?
+        .ok_or_else(|| {
+            log::error!("Could not find database: {}", db_identity.to_hex());
+            NO_SUCH_DATABASE
+        })?;
+
+    let leader = worker_ctx
+        .leader(database.id)
+        .await
+        .map_err(log_and_500)?
+        .ok_or(StatusCode::NOT_FOUND)?;
+    let module = leader.module().await.map_err(log_and_500)?;
```

Comment:
Looks like this could be shared by `call`, `procedure`, `sql_direct`, and `schema`. The `module` has all the necessary info, including `database`.